### ISSUE TITLE
Remove fileeditor dependency from codemirror-extension package

### ISSUE THREAD

**Issue by jasongrout** [jupyterlab/jupyterlab#10352]
tldr; I think the [@jupyterlab/codemirror-extension](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codemirror-extension) should not depend on the [@jupyterlab/fileeditor](https://github.com/jupyterlab/jupyterlab/tree/master/packages/fileeditor) package.

We currently have an awkward dependency relationship, where the [@jupyterlab/codemirror-extension](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codemirror-extension) uses the resource from the [@jupyterlab/fileeditor-extension](https://github.com/jupyterlab/jupyterlab/tree/master/packages/fileeditor-extension) (through a token provided by the [@jupyterlab/fileeditor](https://github.com/jupyterlab/jupyterlab/tree/master/packages/fileeditor) package) so that the codemirror extension can modify full-document file editors when default settings change, etc. This feels wrong to me, in that it is hardcoding specific assumptions about one place where the codemirror editors are being used in the codemirror extension. Instead, I think the codemirror extension should provide an interface for signaling when editor options change, or should separately keep track of *any* editors it creates and modify their settings as appropriate.

(Note that this isn't a circular dependency among plugins. The codemirror-extension package provides the IEditorServices in one plugin, and requires the IEditorTracker resource in another.)

# Background

Here is a short analysis of how the various editor packages are related to each other

## [@jupyterlab/codeeditor](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codeeditor)

defines the abstract code editor interface and provides a wrapper widget

### Exports

* ``CodeEditor``: A namespace for many abstract editor functions
  * `IEditor`: an interface for an editor widget
  * `Factory`: factory function to create an IEditor
* ``IEditorFactoryService``: An interface for creating new inline or full-document editors
* ``IEditorMimeTypeService``: The mime type service for an editor - getting mime type by file path or by language. This happens to be provided by codemirror, but in general perhaps is not needed to be provided by an editor?
* ``IEditorServices``: a Token providing an IEditorFactoryService and an IEditorMimeTypeService
* `CodeEditorWrapper`: A code editor widget wrapping a code editor. As an initialization option, it takes a factory to produce a CodeEditor.IEditor

## [@jupyterlab/codemirror](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codemirror)

imports from [@jupyterlab/codeeditor](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codeeditor):
* `CodeEditor`
* `IEditorFactoryService`
* `IEditorMimeTypeService`
* `IEditorServices`

### Exports

* `CodeMirrorEditor`: Implements the CodeEditor.IEditor interface, i.e., an actual editor widget built on codemirror
* `editorServices`: Implements IEditorServices token interface
* `EditorSyntaxStatus`: A VDomRenderer to show in the status bar the language syntax highlighting
* `ICodeMirror`: A token to access the CodeMirror singleton
* `Mode`

## [@jupyterlab/fileeditor](https://github.com/jupyterlab/jupyterlab/tree/master/packages/fileeditor)

imports from [@jupyterlab/codeeditor](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codeeditor):
* `CodeEditor`
* `CodeEditorWrapper`
* `IEditorMimeTypeService`
* `IEditorServices`

### Exports

* `FileEditor`: a file editor widget, which contains a FileEditorCodeWrapper, which extends CodeEditorWrapper
* `FileEditorFactory`: An ABCWidgetFactory (from docregistry) for FileEditors
* `IEditorTracker`: A Token for a file editor tracker
* `TabSpaceStatus`

## [@jupyterlab/fileeditor-extension](https://github.com/jupyterlab/jupyterlab/tree/master/packages/fileeditor-extension)

imports from [@jupyterlab/codeeditor](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codeeditor):
* `CodeEditor`
* `IEditorServices`

imports from [@jupyterlab/codemirror](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codemirror):
* `CodeMirrorEditor`

imports from [@jupyterlab/fileeditor](https://github.com/jupyterlab/jupyterlab/tree/master/packages/fileeditor):
* `FileEditor`
* `FileEditorFactory`
* `IEditorTracker`
* `TabSpaceStatus`

Provides plugins:
* `IEditorTracker` - the tracker for full-document file editors

## [@jupyterlab/codemirror-extension](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codemirror-extension)

imports from [@jupyterlab/codeeditor](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codeeditor):
* `IEditorServices`

imports from [@jupyterlab/codemirror](https://github.com/jupyterlab/jupyterlab/tree/master/packages/codemirror):
* `editorServices`
* `EditorSyntaxStatus`
* `CodeMirrorEditor`
* `Mode`
* `ICodeMirror`

imports from [@jupyterlab/fileeditor](https://github.com/jupyterlab/jupyterlab/tree/master/packages/fileeditor):
* `IEditorTracker`
* `FileEditor`

Provides plugins:

* `ICodeMirror`: a CodeMirror singleton
* `IEditorServices`: to create new codemirror inline or full-document widgets
* `commands`: provides various commands for codemirror editors. Requires the file editor tracker, which I think is problematic.
* `editorSyntaxStatus`: adds a status bar widget for the syntax highlight status. Requires the file editor tracker, which I think is problematic.

**Comment by jasongrout** [comment]
I think the codemirror extension factory functions should not keep track of editors it creates and automatically modify relevant settings. For one, the defaults already are a layering of various defaults in 
https://github.com/jupyterlab/jupyterlab/blob/4cb624755d6ddf4e59b3993fd9fcebaf255ec5cb/packages/codemirror/src/factory.ts#L22-L51
Second, any editor may have its options changed by whoever created it. If we wanted to automatically change the defaults, we'd need to take into account any modifications of those defaults, which is impossible for the underlying factory to do.

So that leaves us with a way for the factory to signal that the factory defaults have changed, then it is up to whoever is using the factory to apply the new defaults to existing editors, so perhaps the signal is an IChanged signal that gives the old and new defaults.

A much simpler approach would be to keep the existing code for applying defaults (i.e., they do not update in-place, and are only applied at application launch) and just tell the user that they need to refresh if they ever change the codemirror defaults.

**Comment by jasongrout** [comment]
I explored a bit with some of these ideas in https://github.com/jasongrout/jupyterlab/tree/editordefaultsettings (commit  https://github.com/jasongrout/jupyterlab/commit/8a6a27d15994138893eb1c27c1ac04c5d5a44745)

**Comment by fcollonval** [comment]
Closing as fixed by https://github.com/jupyterlab/jupyterlab/pull/13639
